Add localAddress Support in Java Http3 ClientBuilder for Network Interface Selection

kdhDevelop opened the pull request:
This pull request support for specifying a local network interface in the Http3 ClientBuilder. The localAddress feature enables developers to configure which network interface should be used for outgoing communications.

localAddress Implementation : Added support for the localAddress property in the ClientBuilder. Users can now set the network interface to control the source of their network traffic.

The creator interface of the HTTP3SingleConnectionClient has been modified. You must choose whether to modify this part or disable the network interface selection function and use an existing creator.

Additional Notes:
I'm a novice developer. Feedback and suggestions are welcome. Let me know if additional adjustments are needed!
I am not good at English. I am using translator and GPT. Please understand if my English is strange.

ptrd commented on src/main/java/net/luminis/http3/impl/InterfaceBoundDatagramSocketFactory.java:
```diff
@@ -0,0 +1,24 @@
+public class InterfaceBoundDatagramSocketFactory implements DatagramSocketFactory {
+    private final InetAddress address;
+    public InterfaceBoundDatagramSocketFactory(InetAddress address) {
+        this.address = address;
```
whose address? suggestion: use a descriptive name instead off "address"

ptrd commented on src/main/java/net/luminis/http3/impl/InterfaceBoundDatagramSocketFactory.java:
```diff
@@ -0,0 +1,24 @@
+    public InterfaceBoundDatagramSocketFactory(InetAddress address) {
+        this.address = address;
+    }
+    private DatagramSocket bindToLocalAddress(InetAddress address) throws SocketException {
```
same here: use better name than just "address"

ptrd commented on src/main/java/net/luminis/http3/impl/InterfaceBoundDatagramSocketFactory.java:
```diff
@@ -0,0 +1,24 @@
+        return new DatagramSocket(new InetSocketAddress(address, 0));
+    }
+    @Override
+    public DatagramSocket createSocket(InetAddress inetAddress) throws SocketException {
```
and again here: use better name than just "address"

ptrd commented on src/main/java/net/luminis/http3/impl/InterfaceBoundDatagramSocketFactory.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public DatagramSocket createSocket(InetAddress inetAddress) throws SocketException {
+        DatagramSocket datagramSocket = bindToLocalAddress(address);
+        datagramSocket.connect(new InetSocketAddress(inetAddress, 443));
```
hard coded 443??? This is the address of the server the client is connecting to right? Doesn't have to be 443

ptrd commented on src/main/java/net/luminis/http3/impl/InterfaceBoundDatagramSocketFactory.java:
```diff
@@ -0,0 +1,24 @@
+
+public class InterfaceBoundDatagramSocketFactory implements DatagramSocketFactory {
+    private final InetAddress address;
+    public InterfaceBoundDatagramSocketFactory(InetAddress address) {
```
I wonder what happens if someone passes a address that is not local. Why not add a unit test that shows what happens (e.g. which exception is thrown) in such case?

ptrd reviewed (changes requested):
Thanks for the PR. Looks good, except for one error and a few tips in InterfaceBoundDatagramSocketFactory.
Also, you will have to hand over copyright to me, so I can publish it under GPL license. Is that ok with you?

ptrd commented:
Your English is fine btw. ;-)

kdhDevelop commented:
@ptrd 
Thank you for your feedback. I’ve reviewed the message, made the necessary corrections, and included the suggested improvements in InterfaceBoundDatagramSocketFactory.

> whose address? suggestion: use a descriptive name instead off "address"
> same here: use better name than just "address"
> and again here: use better name than just "address"

I renamed address to localAddress as it serves the same function as localAddress in Java’s HttpClient.

> hard coded 443??? This is the address of the server the client is connecting to right? Doesn't have to be 443

Upon review, I found that the line was unnecessary, so I removed it.

> I wonder what happens if someone passes a address that is not local. Why not add a unit test that shows what happens (e.g. which exception is thrown) in such case?

A BindException occurs. I have added the unit test InterfaceBoundDatagramSocketFactoryTest to verify its occurrence.

> Thanks for the PR. Looks good, except for one error and a few tips in InterfaceBoundDatagramSocketFactory. Also, you will have to hand over copyright to me, so I can publish it under GPL license. Is that ok with you?

As for the copyright, I’m fine with transferring it to you for publishing under the GPL license. Let me know if there’s anything else you need from my side.

ptrd commented:
Thanks a lot for your contribution @kdhDevelop !
